action/envelope.go

```diff
@@ -230,10 +239,100 @@ func (elp *Envelope) Serialize() []byte {
 	return byteutil.Must(proto.Marshal(elp.Proto()))
 }
 
-// Hash returns the hash value of SealedEnvelope.
+// Hash returns the hash value of Envelope.
 func (elp *Envelope) Hash() hash.Hash256 {
+	if elp.IsExternal() {
+		return elp.rawHash
+	}
 	return hash.Hash256b(elp.Serialize())
 }
 
 // SetNonce sets the nonce value
 func (elp *Envelope) SetNonce(n uint64) { elp.nonce = n }
+
+// InvalidChainID detects invalid chain ID
+func (elp *Envelope) InvalidChainID() bool {
+	return elp.chainID != 0 && elp.chainID != config.ExternChainID()
+}
+
+// IsExternal returns whether the action is external or not
+func (elp *Envelope) IsExternal() bool {
+	return elp.chainID == config.ExternChainID()
+}
+
+// ValidateExternal validates an external action
+func (elp *Envelope) ValidateExternal(sig []byte) error {
```

Comment:
this function has side effort rather than just validate the action

Reply:
yes, it also computes the hash

Reply:
it actually decodes the tx, validation will be done later in Verify()
change name to DecodeExternal()